compiler/rustc_target/src/asm/avr.rs

```diff
@@ -0,0 +1,128 @@
+use super::{InlineAsmArch, InlineAsmType};
+use rustc_macros::HashStable_Generic;
+use std::fmt;
+
+def_reg_class! {
+    Avr AvrInlineAsmRegClass {
+        reg,
+        reg_xyz,
+    }
+}
+
+impl AvrInlineAsmRegClass {
+    pub fn valid_modifiers(self, _arch: InlineAsmArch) -> &'static [char] {
+        &[]
+    }
+
+    pub fn suggest_class(self, _arch: InlineAsmArch, _ty: InlineAsmType) -> Option<Self> {
+        None
+    }
+
+    pub fn suggest_modifier(
+        self,
+        _arch: InlineAsmArch,
+        _ty: InlineAsmType,
+    ) -> Option<(char, &'static str)> {
+        None
+    }
+
+    pub fn default_modifier(self, _arch: InlineAsmArch) -> Option<(char, &'static str)> {
+        None
+    }
+
+    pub fn supported_types(
+        self,
+        _arch: InlineAsmArch,
+    ) -> &'static [(InlineAsmType, Option<&'static str>)] {
+        match self {
+            Self::reg => types! { _: I8; },
+            Self::reg_xyz => types! { _: I16; },
+        }
+    }
+}
+
+def_regs! {
+    Avr AvrInlineAsmReg AvrInlineAsmRegClass {
+        r0: reg = ["r0"],
+        r1: reg = ["r1"],
+        r2: reg = ["r2"],
+        r3: reg = ["r3"],
+        r4: reg = ["r4"],
+        r5: reg = ["r5"],
+        r6: reg = ["r6"],
+        r7: reg = ["r7"],
+        r8: reg = ["r8"],
+        r9: reg = ["r9"],
+        r10: reg = ["r10"],
+        r11: reg = ["r11"],
+        r12: reg = ["r12"],
+        r13: reg = ["r13"],
+        r14: reg = ["r14"],
+        r15: reg = ["r15"],
+        r16: reg = ["r16"],
+        r17: reg = ["r17"],
+        r18: reg = ["r18"],
+        r19: reg = ["r19"],
+        r20: reg = ["r20"],
+        r21: reg = ["r21"],
+        r22: reg = ["r22"],
+        r23: reg = ["r23"],
+        r24: reg = ["r24"],
+        r25: reg = ["r25"],
+        r26: reg = ["r26", "XL"],
+        r27: reg = ["r27", "XH"],
+        r28: reg = ["r28", "YL"],
+        r29: reg = ["r29", "YH"],
+        r30: reg = ["r30", "ZL"],
+        r31: reg = ["r31", "ZH"],
+
+        X: reg_xyz = ["r27r26", "X"],
+        Y: reg_xyz = ["r29r28", "Y"],
```

Comment:
If you have a look at `lib/Target/AVR/AVRRegisterInfo.cpp` in the LLVM source you will see that `Y` is actually reserved for the stack pointer. Also `r0` and `r1` are reserved.

Reply:
It looks like for `Y` the way to go is adding the `#error` pragma here to keep the user's hands off of it.

I think we'd need some way to name `r0` and `r1`.  For instance, to get the result of a `MUL`:

```
let a = 1u8;
let b = 1u8;
let res: u16;
asm!(
    "mul {}, {}",
    in(reg) a,
    in(reg) b,
    lateout("r1r0") res,
);
```

Is there another architecture that I can look to for an example of how to handle this?

---

(and just to clarify because it confused me for a moment, `Y` is reserved for the frame pointer, the stack pointer has a dedicated register `SPH:SPL`)

Reply:
> I think we'd need some way to name r0 and r1.

Or just leave it to the user to reassemble the value from `r0` and `r1` themselves. In the end it should be optimized the same way by the compiler.

Reply:
Sorry, I wasn't clear.  If `r0` and `r1` are reserved, and thus probably shouldn't be nameable for inline `asm!`, how is the user expected to get the results from them?

Or to put it another way, what should I do with the information that `lib/Target/AVR/AVRRegisterInfo.cpp` reserves these registers if not exclude them from use?

Reply:
> Sorry, I wasn't clear. If r0 and r1 are reserved, and thus probably shouldn't be nameable for inline asm!, how is the user expected to get the results from them?

Both registers `r1` and `r0` should be accessible. Any author of some piece Rust code targeting AVR should be able to name these registers in inline ASM.

The zero register should always be accessible even in inline asm so that the dev can pass `0` to an instruction which only supports registers as inputs, not immediates.

See this avr-gcc wiki doc https://gcc.gnu.org/wiki/avr-gcc#Fixed_Registers which shows that both registers should be available in inline asm.

It is a little bit finicky with LLVM instruction reordering, but it's always like that and I don't see a problem with your approach. You won't be able to say, run a piece of C code (i.e. something that executes `MUL`) and then use inline ASM to reassemble `r1`/`r0` because LLVM inline assembler does not parse/inspect into the raw assembly text (only the constraints/list), so LLVM won't know you're referring to r1/r0 and so will not be able to preserve it during transformation. You would be able to use inline ASM to reassemble r1/r0 from the result of an instruction that is also inside the inline asm block, because the internal contents of any inline asm block would not be reordered. **In summary, in this paragraph all I am trying to say that it is perfectly acceptable and allowable to use `r1`/`r0` in inline ASM, but you cannot rely on their values across inline asm block boundaries**

Reply:
It sounds like there should be a register class just for `r1`/`r0`, they shouldn't be in `reg` because we wouldn't want to pick either of them for "any old register".

> The zero register should always be accessible

Thanks for bringing this up.  I'll need to add to the documentation that if a programmer writes to `r1` in any way that they must restore it to zero before the end of the `asm!` block.

> it is perfectly acceptable and allowable to use r1/r0 in inline ASM, but you cannot rely on their values across inline asm block boundaries.

Does that mean that you shouldn't be able to use them as inputs or outputs to the block?  For instance, would my code sample up thread be expected to work, or does that count as "across block boundaries"?

Reply:
I agree that `r0` and `r1` *should* be usable as operands in `asm!`. However the reality is that LLVM marks these as reserved registers, which causes LLVM's register allocator to ignore these and assume they are never touched by inline assembly. If you try to actually use `r0`/`r1` as operands then things will break in unpredictable ways, which is why we always prevent reserved registers from being used in `asm!`.

The only "correct" way to use `r0` and `r1` in `asm!` is to save their value to another register (which should be properly marked as clobbered) and then restoring these values at the end of the assembly snippet. This is unavoidable until the LLVM backend is fixed to no longer treat these registers as reserved.

Reply:
Source: https://github.com/llvm/llvm-project/blob/27c9e8b45b25614a92539ac6787dbb5670d950b3/llvm/lib/Target/AVR/AVRRegisterInfo.cpp#L53

Reply:
I am having a hard time understanding exactly what LLVM means by reserved.  The only authority I've found is the comment above `getReservedRegs` here: https://github.com/llvm/llvm-project/blob/e356027016c6365b3d8924f54c33e2c63d931492/llvm/include/llvm/CodeGen/TargetRegisterInfo.h#L524-L527

```
  /// A reserved register:
  /// - is not allocatable
  /// - is considered always live
  /// - is ignored by liveness tracking
```

I don't think I know enough to tell if that should be applied to `r0` and `r1` or not.  I'm inclined to trust your judgement there @Amanieu.

> This is unavoidable until the LLVM backend is fixed to no longer treat these registers as reserved.

If this is where we're going, I'm happy to submit this change to LLVM.  @dylanmckay, since I suspect you'll be reviewing the differential, does that change make sense to you?

In the meantime, I'll make `r0` and `r1` error out here with a helpful message.

Reply:
> > This is unavoidable until the LLVM backend is fixed to no longer treat these registers as reserved.

> If this is where we're going, I'm happy to submit this change to LLVM. @dylanmckay, since I suspect you'll be reviewing the differential, does that change make sense to you?

From my point of view, if AVR-GCC compatibility is kept, then the change is definitely good. If not, maybe it's possible but certainly more complicated :+1: